### PR TITLE
Define collector component and remove global references 

### PR DESCRIPTION
**Pull request by gh123man**
### What does this PR do?

This PR defines the component interface for the collector and uses it where applicable. The implementation will be moved into `comp/collector` in a followup PR. 
This PR also removes all global usage of the collector. 

You may choose to review this PR in 2 parts:
1. Define and use the component interface https://github.com/DataDog/datadog-agent/pull/22533/commits/79447a95d7446cf03a64b3abfadfbb3379bb8c1a
2. Remove global references to the collector https://github.com/DataDog/datadog-agent/pull/22533/commits/9617492055035a6b4e134c694e85bc9f19ce7767

This PR also clarifies the conditions where it is expected to have a `nil` collector vs a non-`nil` collector by using `Optional` in the relevant places (such as subcommands)

### Describe how to test/QA your changes

- Test the agent diagnose command still works and runs check diagnosis
- Test that core checks still work
- Test that python checks still work
- Test JMX subcommand still works
- Test agent gui subcommand still works
- Test `agent status` and the collector section still shows up

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: comp/collector/collector/component.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	"github.com/DataDog/datadog-agent/pkg/collector/check"
+	checkid "github.com/DataDog/datadog-agent/pkg/collector/check/id"
+)
+
+// team: agent-metrics-logs
+
+// Component is the component type.
+type Component interface {
+	// Start begins the collector's operation.  The scheduler will not run any checks until this has been called.
+	Start()
```
Comment (gh123man) [inline comp/collector/collector/component.go:20]: I plan to remove start/stop in favor of lifecycle hooks in the next PR

File: comp/collector/component.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+	"github.com/DataDog/datadog-agent/pkg/collector/check"
+	checkid "github.com/DataDog/datadog-agent/pkg/collector/check/id"
+)
+
+// team: agent-metrics-logs
+
+// Component is the component type.
+type Component interface {
+	// Start begins the collector's operation.  The scheduler will not run any checks until this has been called.
+	Start()
```
Comment (ogaca-dd) [inline comp/collector/component.go]: Can you add a TODO to remove `Start` and `Stop` as there is a on going task to remove these kind of methods for components?

**Review by ogaca-dd — APPROVED** [review]
LGTM

File: cmd/cluster-agent-cloudfoundry/subcommands/run/command.go [hunk 3]
```diff
@@ -139,12 +143,11 @@ func run(log log.Component, taggerComp tagger.Component, demultiplexer demultipl
...
 	// The Autoconfig instance setup happens in the workloadmeta start hook
-	// create and setup the Collector and others.
-	common.LoadComponents(demultiplexer, secretResolver, wmeta, pkgconfig.Datadog.GetString("confd_path"))
+	common.LoadComponents(secretResolver, wmeta, pkgconfig.Datadog.GetString("confd_path"))
 
 	// Set up check collector
-	common.AC.AddScheduler("check", collector.InitCheckScheduler(common.Coll, demultiplexer), true)
-	common.Coll.Start()
+	common.AC.AddScheduler("check", pkgcollector.InitCheckScheduler(optional.NewOption[pkgcollector.Collector](collector), demultiplexer), true)
+	collector.Start()
```
Comment (ogaca-dd) [inline cmd/cluster-agent-cloudfoundry/subcommands/run/command.go:151]: You may miss to initialize diagnose package.

File: comp/collector/collector/collectorimpl/collector.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+func Module() fxutil.Module {
+	return fxutil.Component(
+		fx.Provide(newCollector),
+		fx.Provide(func(collector collector.Component) optional.Option[collector.Component] {
+			return optional.NewOption(collector)
+		}),
+	)
+}
+
+func newCollector(deps dependencies) collector.Component {
```
Comment (GustavoCaso) [inline comp/collector/collector/collectorimpl/collector.go:46]: Could we move the status provider for collector in the collector component? 

https://github.com/DataDog/datadog-agent/blob/63f3c76fd79585e066f53fd60c188fdfc247de33/cmd/agent/subcommands/run/command.go#L307

We can inject using FX value groups.


Here is an example of a component that export a status provider
https://github.com/DataDog/datadog-agent/blob/15442f6ed546c30ae5c30535a21ccce097f584e9/comp/metadata/inventoryagent/inventoryagentimpl/inventoryagent.go#L125

That way, FX takes care of injecting the dependencies in the status component.

If something is not clear, do not hesitate to reach 😄

File: cmd/cluster-agent-cloudfoundry/subcommands/run/command.go (hunk 3 above)
Comment (gh123man) [reply cmd/cluster-agent-cloudfoundry/subcommands/run/command.go:151]: I'll double check this one thanks for the heads up! It was difficult to determine all the places the diagnose package needed to be initialized. Glad to hear you are already working on this though!

File: comp/collector/collector/collectorimpl/collector.go (hunk 4 above)
Comment (gh123man) [reply comp/collector/collector/collectorimpl/collector.go]: @GustavoCaso Thanks for pointing this out! I moved the status provider into the component implementation :)

**Review by rayz — APPROVED** [review]
LGTM

**Review by jennchenn — APPROVED** [review]

**Review by sarah-witt — APPROVED** [review]

**Pull request merged**